src/v/ssx/async_algorithm.h

```diff
@@ -299,4 +300,76 @@ ss::future<> async_for_each_counter(
       detail::ref_counter{counter.count}, begin, end, std::move(f));
 }
 
+/**
+ * @brief Call f on every element, yielding occasionally and accepting
+ * an externally provided counter for yield control.
+ *
+ * This is equivalent to std::for_each, except that the computational
+ * loop yields every Traits::interval (default 100) iterations in order
+ * to avoid reactor stalls. The returned future resolves when all elements
+ * have been processed.
+ *
+ * This behaves similarly to async_for_each except that the counter used to
+ * track how much work as been done since the last attempted yield is passed
+ * in by the caller. This allows use in more complex cases such as nested loops
+ * where the inner loop may itself not do sufficient work to ever trigger the
+ * yield condition, even though the total amount of work done across all
+ iterations
+ * is very high.
+ *
+ * Use case:
+ *
+ * Replace something like:
+ *
+ * for (... outer loop ...) {
+ *   for (... inner loop ...) {
+ *      f(elem);
+ *   }
+ * }
+ * with:
+ *
+ * async_counter counter;
+ * for (... outer loop ...) {
+ *    co_await async_for_each(counter, begin, end, f);
+ * }
+ *
+ * The counter is taken by reference and must live at least until the
+ * returned future resolves: this usually trivial when the caller is a
+ * coroutine but may require some care when continuation style is used.
+ *
+ * The function is taken by value.
+ *
+ * Until the returned future resolves, the container's begin and end iterators,
+ * as they were when called, must remain valid, and the container itself must
+ * remain live if it was passed via lvalue reference.
+ *
+ * @param counter the counter object to use, may be reused across invocations
+ * @param container universal reference to container
+ * @param f the function to call on each element
+ * @return ss::future<> a future which resolves when all elements have been
+ * processed
+ */
+template<typename Traits = async_algo_traits, typename Fn, typename Container>
+requires requires(Container c, Fn fn) {
+    { ss::futurize_invoke(fn, *std::begin(c)) } -> std::same_as<ss::future<>>;
+    std::end(c);
+}
+ss::future<>
+async_for_each_counter(async_counter& counter, Container&& container, Fn f) {
+    if constexpr (std::is_lvalue_reference_v<decltype(container)>) {
+        return detail::async_for_each_fast<Traits>(
+          counter, std::begin(container), std::end(container), std::move(f));
+    } else {
+        return ss::do_with(
+          std::move(container),
+          [&counter, f = std::move(f)](Container& container) {
```

Comment:
@bashtanov It seems like a footgun having different lifetime semantics depending on the value category. Why not pick one of the two and let the caller deal with it? For example it's very common for a caller to introduce the `do_with` if needed (or put some data on the stack and use co_await). Maybe i'm missing something here? cc @mmaslankaprv

Reply:
I think it's quite natural:
* if a caller provides an lvalue they do not expect us to move or copy from it, so it remains for them to hold it;
* if a caller provides an rvalue they expect us to take ownership of the data, including storing it for as long as necessary.
I even thought a `do_with` flavour with this logic would be quite useful. I would say it's more natural than silently making a copy.

> Why not pick one of the two and let the caller deal with it?

To reduce boilerplate for the convenience of the callers.

Reply:
I don't disagree with your analysis, but I think the thing that is being discounted here is consistency in behavior. AFAICT this is not consistent with most seastar/ssx APIs, right?

Reply:
Thanks, your point makes sense to me. Luckily all existing use of these two functions (there's one more above) will remain valid if we remove the `do_with`, as they all are co-awaited straight away. I'll open a thread in #cpplang tomorrow to gather more opinions.